backend/src/integration/fitness/providers/fitbit.provider.ts

```diff
@@ -0,0 +1,262 @@
+      description:
+        'Access Activities, Steps and further Data from Fitbit. This supports all devices supported by the Fitbit App, including the Google Pixel Watch',
+      status: this.userStatus,
+    };
+  }
+
+  /**
+   * Exchanges a code recieved after an authorized to an access token
+   * and persists the client
+   *
+   * @param user internal id of the user
+   * @param code received authorization code
+   * @returns api credentials
+   */
+  public async getAccessTokenFromCode(
+    user: string,
+    code: string,
+  ): Promise<FitbitCredentials> {
+    const searchParams = new URLSearchParams();
+    searchParams.append('client_id', this.client_id);
+    searchParams.append('code', code);
+    searchParams.append('grant_type', 'authorization_code');
+
+    // Retrieve the access token from the server
+    const response = await fetch(`${this.FITBIT_API}/oauth2/token`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `Basic ${Buffer.from(
+          `${this.client_id}:${this.client_secret}`,
+        ).toString('base64')}`,
+      },
+      body: searchParams.toString(),
+    });
+
+    if (!response.ok) {
+      console.log(await response.text());
+      throw new Error('Invalid response from FitBit');
+    }
+
+    const { access_token, refresh_token, expires_in, user_id } =
+      await response.json();
+
+    this.userStatus = 'enabled';
+
+    await this.fitnessRepository.createProvider({
+      type: 'fitbit',
+      providerUserId: user_id,
+      accessToken: access_token,
+      accessTokenExpires: dayjs().add(expires_in, 'seconds').toDate(),
+      refreshToken: refresh_token,
+      enabled: true,
+      owner: {
+        connect: {
+          id: user,
+        },
+      },
+    });
+
+    return { accessToken: access_token, userId: user_id };
+  }
+
+  /**
+   * Retrieves a new access token for a configured datasource.
+   * The datasource needs to be configured otherwise this will fail.
+   *
+   * @param user internal id of the user
+   * @returns api credentials
+   */
+  private async getAccessToken(user: string): Promise<FitbitCredentials> {
+    // First check, whether the access token is still cached
+    const credentialsCache = this.getCredentialsFromCache(user);
+    if (credentialsCache) {
+      return credentialsCache;
+    }
+
+    const credentials = await this.fitnessRepository.getProviderForUserById(
+      user,
+      'fitbit',
+    );
+
+    if (!credentials) throw new Error('No credentials found for user');
+
+    // Construct the URL
+    const searchParams = new URLSearchParams();
+    searchParams.append('client_id', this.client_id);
+    searchParams.append('refresh_token', credentials.refreshToken);
+    searchParams.append('grant_type', 'refresh_token');
+
+    // Retrieve the access token from the server
+    const response = await fetch(`${this.FITBIT_API}/oauth2/token`, {
+      method: 'POST',
```

Comment:
We could argue about extracting the save action, happening in the "setTimeout" block in 160f, but I don't think this would bring much benefit, as we need to pass many parameters. As for calling the API: I don't think there is any useful abstraction.